Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp

```diff
@@ -734,30 +1073,201 @@ void ProgramShaderCache::CreateHeader()
+
+        // The ubershader may already exist if shader caching is enabled.
+        if (!success || ubershaders.find(uid) != ubershaders.end())
+          return;
+
+        PCacheEntry& entry = ubershaders[uid];
+        entry.in_cache = false;
+        entry.pending = true;
+
+        // Multi-context path?
+        if (s_async_compiler)
+        {
+          s_async_compiler->QueueWorkItem(
+              s_async_compiler->CreateWorkItem<UberShaderCompileWorkItem>(uid));
+          return;
+        }
+
+        ShaderCode vcode = UberShader::GenVertexShader(APIType::OpenGL, uid.vuid.GetUidData());
+        ShaderCode pcode = UberShader::GenPixelShader(APIType::OpenGL, uid.puid.GetUidData());
+        ShaderCode gcode;
+        if (g_ActiveConfig.backend_info.bSupportsGeometryShaders &&
+            !uid.guid.GetUidData()->IsPassthrough())
+        {
+          GenerateGeometryShaderCode(APIType::OpenGL, uid.guid.GetUidData());
+        }
+
+        // Always background compile, even when it's not supported.
+        // This way hopefully the driver can still compile the shaders in parallel.
+        if (!CompileShader(entry.shader, vcode.GetBuffer(), pcode.GetBuffer(), gcode.GetBuffer(),
+                           true))
+        {
+          // Stop compiling shaders if any of them fail, no point continuing.
+          success = false;
+          return;
+        }
+      });
+    });
+  });
+
+  if (s_async_compiler)
+  {
+    s_async_compiler->WaitUntilCompletion([](size_t completed, size_t total) {
+      Host_UpdateProgressDialog(GetStringT("Compiling shaders...").c_str(),
+                                static_cast<int>(completed), static_cast<int>(total));
+    });
+    s_async_compiler->RetrieveWorkItems();
+    Host_UpdateProgressDialog("", -1, -1);
+  }
 
-  PCacheEntry entry;
-  entry.in_cache = 1;
-  entry.shader.glprogid = glCreateProgram();
-  glProgramBinary(entry.shader.glprogid, *prog_format, binary, binary_size);
+  // Ensure all the ubershaders are compiled before booting.
+  if (success)
+  {
+    for (auto& it : ubershaders)
+    {
+      PCacheEntry* entry = &it.second;
+      if (entry->pending)
+      {
+        if (!entry->shader.glprogid || (!s_async_compiler && !entry->FinishParallelCompile()))
+        {
+          // If any ubershaders failed to compile, throw out all of them.
+          // Otherwise, we'll potentially use an invalid/non-existent ubershader.
+          success = false;
+          break;
+        }
+      }
+    }
+  }
 
-  GLint success;
-  glGetProgramiv(entry.shader.glprogid, GL_LINK_STATUS, &success);
+  if (!success)
+  {
+    PanicAlert("One or more ubershaders failed to compile. Disabling ubershaders.");
+    for (auto& it : ubershaders)
+      it.second.Destroy();
+    ubershaders.clear();
+  }
+}
 
-  if (success)
+bool ProgramShaderCache::SharedContextAsyncShaderCompiler::WorkerThreadInitMainThread(void** param)
+{
+  auto ctx = GLInterface->CreateSharedContext();
```

Comment:
This only appears to be implemented for EGL and WGL GLInterfaces - this PanicAlert is always triggered on GLX, and I have no idea how/if it's working on Mac.